test/Main.hs

```diff
@@ -227,24 +227,76 @@ assertEqual preface expected actual eq =
 
 ----------------------------------------
 
+sqlGenInts :: Int32 -> SQL
+sqlGenInts n = smconcat
+  [ "WITH RECURSIVE ints(n) AS"
+  , "( VALUES (1) UNION ALL SELECT n+1 FROM ints WHERE n <" <?> n
+  , ") SELECT n FROM ints"
+  ]
+
+cursorTest :: TestData -> Test
+cursorTest td = testGroup "Cursors"
+  [ basicCursorWorks
+  , scrollableCursorWorks
+  , withHoldCursorWorks
+  ]
+  where
+    basicCursorWorks = testCase "Basic cursor works" $ do
+      runTestEnv td def . withCursor "ints" mempty (sqlGenInts 5) $ \cursor -> do
+        xs <- (`fix` []) $ \loop acc -> cursorFetch cursor CD_Next >>= \case
+          0 -> return $ reverse acc
+          1 -> do
+            (n::Int32) <- fetchOne runIdentity
+            loop $ n : acc
+          n -> error $ "Unexpected number of rows: " ++ show n
+        assertEqual "Data fetched correctly" [1..5] xs (==)
+
+    scrollableCursorWorks = testCase "Cursor declared as SCROLL works" $ do
+      runTestEnv td def . withCursor "ints" scroll (sqlGenInts 10) $ \cursor -> do
+        checkMove cursor CD_Next         1
+        checkMove cursor CD_Prior        0
+        checkMove cursor CD_First        1
+        checkMove cursor CD_Last         1
+        checkMove cursor CD_Backward_All 9
+        checkMove cursor CD_Forward_All  10
+        checkMove cursor (CD_Absolute 0) 0
+        checkMove cursor (CD_Relative 0) 0
+        checkMove cursor (CD_Forward 5)  5
+        checkMove cursor (CD_Backward 5) 4
+
+        cursorFetch_ cursor CD_Forward_All
+        xs1::[Int32] <- fetchMany runIdentity
+        assertEqual "xs1 is correct" [1..10] xs1 (==)
+        cursorFetch_ cursor CD_Backward_All
+        xs2::[Int32] <- fetchMany runIdentity
+        assertEqual "xs2 is correct" (reverse [1..10]) xs2 (==)
```

Comment:
Looks like there's only one place in this file where we don't use `assertEqual` with `(==)`. Would make more sense to use the version with an `Eq` constraint the default and rename the current one to e.g. `assertEqual'`.

Reply:
Yeah, we can do that. Call sites also mess up the `expected` with `given` case in some places.